pkg/retriever/graphsyncretriever.go

```diff
@@ -91,9 +74,16 @@ type retrievalResult struct {
 // retrieval handles state on a per-retrieval (across multiple candidates) basis
 type graphsyncRetrieval struct {
 	*GraphSyncRetriever
-	ctx            context.Context
-	request        types.RetrievalRequest
-	eventsCallback func(types.RetrievalEvent)
+	ctx                 context.Context
+	request             types.RetrievalRequest
+	eventsCallback      func(types.RetrievalEvent)
+	candidateMetadata   map[peer.ID]*metadata.GraphsyncFilecoinV1
```

Comment:
Might be worth considering whether we can remove the pointer here too.
In the current code it's not inconceivable that getting the metadata from `candidate.Metadata.Get` may yield a `nil` (at least locally here we have no guarantees, likely it's perfectly safe upstream because why are we getting candidates otherwise? But I think we should treat it with caution since this is a discrete component). Then `metadataCompare` launches ahead assuming that the pointers it has are non-`nil`, without any safety.

I think I'd try removing this pointer type here, safely attempting to resolve the pointer when we set new ones, then just using values from here on. Then at least if you can't resolve the pointer from the candidate then you can skip it entirely.

Reply:
I like the discrete component perspective. Upstream, the candidate finder is ensuring the candidates have a metadata, but covering our bases here will ensure we'll never panic on a nil `metadata.Protocol`.

Reply:
Ah, nevermind, only the `metdata.GraphsyncFilecoinV1` pointer receiver implements `metadata.Protocol`.

Reply:
> I think I'd try removing this pointer type here, safely attempting to resolve the pointer when we set new ones

Ah okay, you already mention this. Yeah, I can resolve the pointer and then use the values throughout.

Reply:
So this makes me want to check that the graphsync metadata we're grabbing and then casting from a candidate for the first time is okay. If we've seen the peer ID before, we can grab the candidate's metadata and check if it's more preferable. If the cast was bad, we continue.

```golang
// Grab the current candidate's metadata
candidateMetadata, ok := candidate.Metadata.Get(multicodec.TransportGraphsyncFilecoinv1).(*metadata.GraphsyncFilecoinV1)
...
// Don't start a new retrieval if we've seen this candidate before, but update the metadata if it's more favorable
if seenCandidate {
	if !ok {
		continue
	}
        ...
}
...
```
This is okay because we assume we were able to save some metadata when we first encountered the peer ID.

In the case that we haven't seen the peer ID before _and_ the cast to the graphsync metadata was not okay, we could create an empty graphsync metadata using the CID in the request to populate the PieceCID property, allowing default falses for the Fast and Verified properties.
```golang
// Grab the current candidate's metadata
candidateMetadata, ok := candidate.Metadata.Get(multicodec.TransportGraphsyncFilecoinv1).(*metadata.GraphsyncFilecoinV1)
...
// Track the candidate metadata, creating a new one if needed
if !ok {
	candidateMetadata = &metadata.GraphsyncFilecoinV1{PieceCID: r.request.Cid}
}
r.candidateMetdataLk.Lock()
r.candidateMetadata[candidate.MinerPeer.ID] = *candidateMetadata
r.candidateMetdataLk.Unlock()
...
```
I'm not sure if this is the correct way to use a graphsync metadata though. Thoughts?

Reply:
Looking at https://github.com/ipni/go-libipni/tree/main/metadata, I believe that it's impossible to get a metadata for `TransportGraphsyncFilecoinv1` that isn't a `*GraphsyncFilecoinV1`. So I think you're good on that first check.

When the metadata is decoded, it's done according to this schema: https://github.com/ipni/go-libipni/blob/main/metadata/graphsync_filecoinv1.ipldsch - notably there's no optional fields in that, and likewise there should be no pointers in that type. So if you did need to instantiate a new one manually, then just the PieceCID _not_ being a `cid.Undef` (zero value) would be enough to get a properly formed one I think. But as I said, I don't think you will need this `if`.

Sad that its only the pointer form that implements the interface, no good reason for `ID()` to have a pointer receiver, I guess it's only like that because of all the marshalling code needs pointer forms so it's there for consistency.